storage/stores.go

```diff
@@ -225,3 +246,75 @@ func (ls *Stores) RangeLookup(key roachpb.RKey, _ *roachpb.RangeDescriptor, cons
 	}
 	return br.Responses[0].GetInner().(*roachpb.RangeLookupResponse).Ranges, nil
 }
+
+// ReadBootstrapInfo implements the gossip.Persistence interface. Read
+// attempts to read gossip bootstrap info from every known store and
+// finds the most recent from all stores to initialize the bootstrap
+// info argument. Returns nil if no stores contain gossip bootstrap
+// metadata or an error on any issues reading data from the stores.
+func (ls *Stores) ReadBootstrapInfo(bi *gossip.BootstrapInfo) error {
+	ls.biLatestTS = roachpb.ZeroTimestamp
+	for _, s := range ls.storeMap {
+		value, _, err := engine.MVCCGet(s.engine, keys.StoreGossipKey(), ls.clock.Now(), true, nil)
+		if err != nil {
+			return err
+		}
+		if value != nil && ls.biLatestTS.Less(value.Timestamp) {
+			var storeBI gossip.BootstrapInfo
+			if err := value.GetProto(&storeBI); err != nil {
+				return err
+			}
+			ls.biLatestTS = value.Timestamp
+			ls.gossipBI = storeBI
+		}
+	}
+	*bi = copyBootstrapInfo(ls.gossipBI)
+	return ls.updateAllBootstrapInfos()
+}
+
+// WriteBootstrapInfo implements the gossip.Persistence
+// interface. Write persists the supplied bootstrap info to every
+// known store. Returns nil on success; otherwise returns first error
+// encountered writing to the stores.
+func (ls *Stores) WriteBootstrapInfo(bi gossip.BootstrapInfo) error {
+	ls.biLatestTS = ls.clock.Now()
+	ls.gossipBI = copyBootstrapInfo(bi)
+	for _, s := range ls.storeMap {
+		if err := engine.MVCCPutProto(s.engine, nil, keys.StoreGossipKey(), ls.biLatestTS, nil, &ls.gossipBI); err != nil {
+			return err
+		}
+		log.Infof("wrote gossip bootstrap info to %s", s)
+	}
+	return nil
+}
+
+// updateAllBootstrapInfos cycles through all stores and ensures that
+// the most recent gossip bootstrap information is written to each.
+func (ls *Stores) updateAllBootstrapInfos() error {
+	if ls.biLatestTS.Equal(roachpb.ZeroTimestamp) {
+		return nil
+	}
+	now := ls.clock.Now()
+	for _, s := range ls.storeMap {
+		value, _, err := engine.MVCCGet(s.engine, keys.StoreGossipKey(), now, true, nil)
+		if err != nil {
+			return err
+		}
+		if value == nil || value.Timestamp.Less(ls.biLatestTS) {
+			if err := engine.MVCCPutProto(s.engine, nil, keys.StoreGossipKey(), ls.biLatestTS, nil, &ls.gossipBI); err != nil {
```

Comment:
We don't currently use the MVCC timestamp field for anything that's not updated as a part of the transaction. It would be more consistent with everything else to write these at `ZeroTimestamp` and store a timestamp in the `BootstrapInfo` itself (if we need to - the timestamp comparison looks like a premature optimization to me and we could just blindly overwrite).